missing config leads to crash until you set the record length

This was dastard 0.3.4pre2 (git commit 4521992)

On the DR-TES flight computer, I can reproduce crashes with the following steps:

1) delete `.dastard` directory
2) start dastard
3) connect to ABACO (UDP) source (with 32 channels in this case but I don't know that matters)
4) turn on a auto trigger with delay 0 (for 1 channel)
5) after a short time, dastard crashes with error message below
6) restart dastard
7) connect to ABACO (UDP) source
8) after a short time, dastard crashes. Same error message, just different goroutine number

I then tried the same above steps, but with all 32 channels for the auto trigger. This time dastard starts to allocate huge amounts of memory after step 4.

If, after step 3, I set the record length to a non-zero value, I don't get the crash. Similarly, if I edit the `config.yaml` file to make `nsamples` nonzero, I don't get the crash.

I'm not using `dastardcommander` for these tests, I'm using my own script to command dastard, that I'm writig for dr-tes. I'll attach that script.

Specifically, I run it like:
```
drtes.py udp 191.168.14.100 # to connect to UDP
drtes.py start_noise 1 # to start auto trigger
```

Error message:
```
This is DASTARD version 0.3.4pre2 (git commit 4521992)
Logging problems       to /root/.dastard/logs/problems.log
Logging client updates to /root/.dastard/logs/updates.log

2024/08/23 13:49:37 Dastard config file: /root/.dastard/config.yaml
2024/08/23 13:49:52 New client connection established
2024/08/23 13:49:52 Starting data source named ABACOSOURCE
Sample rate for chan [4096-4127] 46875 /sec determined from  198 packets: Δt=0.050432 sec, Δserial=197, and 12.000 samp/packet
2024/08/23 13:50:52 New client connection established
2024/08/23 13:50:59 New client connection established
Panic! internal buffersChan full, len 100, capacity 100
panic: internal buffersChan full, len 100, capacity 100

goroutine 14 [running]:
github.com/usnistgov/dastard.(*AbacoSource).readerMainLoop(0xc0000e1188)
	/dastard/abaco.go:1071 +0xdf7
created by github.com/usnistgov/dastard.(*AbacoSource).StartRun in goroutine 11
	/dastard/abaco.go:951 +0xd6
```
[drcom.txt](https://github.com/user-attachments/files/16729950/drcom.txt)

--- a/VERSIONS.md
+++ b/VERSIONS.md
@@ -1,9 +1,10 @@
 ## DASTARD Versions
 
-**0.3.4** May 2024-
+**0.3.4** August 23, 2024
 * Update all package dependencies (issue 347).
 * Send which types of files are active in the report about writing data (issue 343).
 * Rename bad Makefile variable `LDFLAGS` -> `GOLINKFLAGS` (issue 350).
+* When no config file exists, set `Nsamples` to a nonzero value, preventing some crashes (issues 355, 356).
 
 **0.3.3** May 28, 2024
 * Make publisher more responsive: don't block data processors waiting on disk flush.

--- a/global_config.go
+++ b/global_config.go
@@ -35,7 +35,7 @@ type BuildInfo struct {
 
 // Build is a global holding compile-time information about the build
 var Build = BuildInfo{
-	Version: "0.3.4pre2",
+	Version: "0.3.4",
 	Githash: "no git hash computed",
 	Date:    "no build date computed",
 }

--- a/rpc_server.go
+++ b/rpc_server.go
@@ -305,15 +305,18 @@ type SizeObject struct {
 // ConfigurePulseLengths is the RPC-callable service to change pulse record sizes.
 func (s *SourceControl) ConfigurePulseLengths(sizes SizeObject, reply *bool) error {
 	*reply = false // handle the case that sizes fails the validation tests and we return early
-	UpdateLogger.Printf("ConfigurePulseLengths: %d samples (%d pre)\n", sizes.Nsamp, sizes.Npre)
 	if !s.isSourceActive {
 		return fmt.Errorf("no source is active")
 	}
+	if sizes.Nsamp <= 0 || sizes.Npre <= 0 {
+		return fmt.Errorf("cannot ConfigurePulseLengths with non-positive values: %v", sizes)
+	}
+	UpdateLogger.Printf("ConfigurePulseLengths: %d samples (%d pre)\n", sizes.Nsamp, sizes.Npre)
 	if s.status.Npresamp == sizes.Npre && s.status.Nsamples == sizes.Nsamp {
 		return nil // no change requested
 	}
 	if s.ActiveSource.WritingIsActive() {
-		return fmt.Errorf("Stop writing before changing record lengths")
+		return fmt.Errorf("stop writing before changing record lengths")
 	}
 
 	f := func() {

--- a/rpc_server.go
+++ b/rpc_server.go
@@ -744,6 +744,10 @@ func (s *SourceControl) StoreRawDataBlock(N int, reply *string) error {
 // (The intention is that block=true in normal operation, but false for certain tests.)
 func RunRPCServer(portrpc int, block bool) {
 	// Set up objects to handle remote calls
+
+	// Set up source control. Later, we will try to use `viper.UnmarshallKey("status"...)' to fill
+	// it with stored values from ~/.dastard/config.yaml. But in the absence of stored values, we
+	// need to set some reasonable defaults, including non-zero values of Nsamples and Npresamples.
 	sourceControl := NewSourceControl()
 	defer sourceControl.lancero.Delete()
 	sourceControl.clientUpdates = clientMessageChan

--- a/rpc_server.go
+++ b/rpc_server.go
@@ -814,6 +814,17 @@ func RunRPCServer(portrpc int, block bool) {
 	}
 
 	err = viper.UnmarshalKey("status", &sourceControl.status)
+	// Set some defaults that won't cause problems down the line.
+	status := &sourceControl.status
+	if status.Npresamp <= 0 {
+		status.Npresamp = 400
+	}
+	if status.Nsamples <= status.Npresamp {
+		status.Nsamples = 2 * status.Npresamp
+	}
+	if status.SamplePeriod <= 0 {
+		status.SamplePeriod = 10 * time.Microsecond
+	}
 	sourceControl.status.Running = false
 	sourceControl.ActiveSource = sourceControl.triangle
 	sourceControl.isSourceActive = false

--- a/rpc_server_test.go
+++ b/rpc_server_test.go
@@ -69,15 +69,6 @@ func TestServer(t *testing.T) {
 	if now := viper.Get("currenttime"); now == nil {
 		t.Errorf("viper.Get(\"currenttime\") returns nil")
 	}
-	sourceControl := NewSourceControl()
-	if err = viper.UnmarshalKey("status", &sourceControl.status); err != nil {
-		t.Errorf("viper.UnmarshalKey(\"status\") returns nil")
-	}
-	if sourceControl.status.Nsamples == 0 {
-		t.Errorf("viper[status] has Nsamples=0")
-	} else if sourceControl.status.Npresamp == 0 {
-		t.Errorf("viper[status] has Npresamples=0")
-	}
 
 	var okay bool
 

--- a/rpc_server_test.go
+++ b/rpc_server_test.go
@@ -399,16 +399,17 @@ func verifyConfigFile(path, filename string) error {
 		}
 	}
 
-	// Create an empty file path/filename, if it doesn't exist.
+	// If it exists, delete it. Then create an empty file path/filename.
 	fullname := filepath.Join(path, filename)
 	_, err = os.Stat(fullname)
-	if os.IsNotExist(err) {
-		f, err := os.OpenFile(fullname, os.O_WRONLY|os.O_CREATE, 0664)
-		if err != nil {
-			return err
-		}
-		f.Close()
+	if !os.IsNotExist(err) {
+		os.Remove(fullname)
 	}
+	f, err := os.OpenFile(fullname, os.O_WRONLY|os.O_CREATE, 0664)
+	if err != nil {
+		return err
+	}
+	f.Close()
 	return nil
 }
 

--- a/rpc_server_test.go
+++ b/rpc_server_test.go
@@ -69,6 +69,15 @@ func TestServer(t *testing.T) {
 	if now := viper.Get("currenttime"); now == nil {
 		t.Errorf("viper.Get(\"currenttime\") returns nil")
 	}
+	sourceControl := NewSourceControl()
+	if err = viper.UnmarshalKey("status", &sourceControl.status); err != nil {
+		t.Errorf("viper.UnmarshalKey(\"status\") returns nil")
+	}
+	if sourceControl.status.Nsamples == 0 {
+		t.Errorf("viper[status] has Nsamples=0")
+	} else if sourceControl.status.Npresamp == 0 {
+		t.Errorf("viper[status] has Npresamples=0")
+	}
 
 	var okay bool
 
@@ -566,6 +575,13 @@ waitingforfile:
 		if data[i] != data[0] {
 			t.Errorf("npz file firstFrameIndex vector not all alike: %v", data)
 		}
+	}
+}
+
+func TestNoConfigFile(t *testing.T) {
+	// Find config file, creating it if needed, and read it.
+	if err := setupViper(); err != nil {
+		panic(err)
 	}
 }
 

